File: app/src/androidTest/java/com/nobrain/samples/daumeimagesearch/network/SearchApiTest.java
package com.nobrain.samples.daumeimagesearch.network;

import android.support.test.runner.AndroidJUnit4;

import com.nobrain.samples.daumeimagesearch.domain.search.SearchChannel;
import com.nobrain.samples.daumeimagesearch.network.retrofit.RetrofitCreator;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import rx.observers.TestSubscriber;

import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;

@RunWith(AndroidJUnit4.class)
public class SearchApiTest {

    private SearchApi searchApi;

    @Before
    public void setUp() throws Exception {
        searchApi = new SearchApi(RetrofitCreator.createRetrofit());
    }

    @Test
    public void testSearchText() throws Exception {
        {
            TestSubscriber<SearchChannel> subscriber = TestSubscriber.create();
            searchApi.searchText("", 1)
                    .subscribe(subscriber);

            List<Throwable> onErrorEvents = subscriber.getOnErrorEvents();
            assertThat(onErrorEvents.size(), is(greaterThan(0)));
        }

        {
            TestSubscriber<SearchChannel> subscriber = TestSubscriber.create();
            searchApi.searchText("kakao", 1)
                    .subscribe(subscriber);

            subscriber.assertCompleted();
            subscriber.assertNoErrors();
            subscriber.assertValueCount(1);
        }
    }
}

File: app/src/androidTest/java/com/nobrain/samples/daumeimagesearch/network/SearchApiTest.kt
package com.nobrain.samples.daumeimagesearch.network

import android.support.test.runner.AndroidJUnit4
import com.nobrain.samples.daumeimagesearch.domain.search.SearchChannel
import com.nobrain.samples.daumeimagesearch.network.retrofit.RetrofitCreator
import org.hamcrest.Matchers.greaterThan
import org.hamcrest.core.Is.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import rx.observers.TestSubscriber

@RunWith(AndroidJUnit4::class)
class SearchApiTest {

    private var searchApi: SearchApi? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        searchApi = SearchApi(RetrofitCreator.createRetrofit())
    }

    @Test
    @Throws(Exception::class)
    fun testSearchText() {
        run {
            val subscriber = TestSubscriber.create<SearchChannel>()
            searchApi!!.searchText("", 1).subscribe(subscriber)

            val onErrorEvents = subscriber.onErrorEvents
            assertThat(onErrorEvents.size, `is`(greaterThan(0)))
        }

        run {
            val subscriber = TestSubscriber.create<SearchChannel>()
            searchApi!!.searchText("kakao", 1).subscribe(subscriber)

            subscriber.assertCompleted()
            subscriber.assertNoErrors()
            subscriber.assertValueCount(1)
        }
    }
}